post.md

```diff
@@ -0,0 +1,427 @@
+```
+
+The server action implementation:
+
+```ts
+"use server";
+
+export async function deleteSomethingWithResponse(
+  _prevState: any,
+  params: FormData
+) {
+  const id = params.get("id");
+  const type = params.get("type");
+
+  console.log(`[${type}] Deleting ${id}...`);
+
+  // Simulate a long-running operation
+  await new Promise((resolve) => setTimeout(resolve, 2000));
+
+  const newId = crypto.getRandomValues(new Uint32Array(1))[0].toString();
+
+  return { status: "success", message: `Deleted ${id}!`, newId };
+}
+```
+
+Note the return value, this will be assigned to `formState` in the client component by `useFormState` hook once the submission is complete. It contains the `status` of the server action, a `message` and a random `newId` that we are going to render in the UI.
+
+We need to import `useFormState` from `react-dom`:
+
+```tsx
+// @ts-ignore
+import { experimental_useFormState as useFormState } from "react-dom";
+```
+
+We need to pass our server action to the `useFormState` hook:
+
+```tsx
+const [formState, formAction] = useFormState(deleteSomethingWithResponse, {});
+```
+
+The `formAction` function is used as the form's action attribute instead of using the server action directly. The `formState` object contains the response from the server, we are going to use the `status` to close the dialog and `newId` to update the UI.
+
+Both of these actions will take place inside a `useEffect` when the form submission is complete. The `useEffect` will run when the `status` or `newId` change which will happen when the form submission is complete.
+
+```tsx
+useEffect(() => {
+  if (status === "success") {
+    setId(newId);
+    close();
+  }
+}, [status, newId, close]);
+```
+
+If the response from the server contains `status: "success"` we update the `id` state and close the modal.
+
+Now our UI is updated as expected, the modal closes and the new id is rendered in the UI.
+
+### Server action signature difference when using `useFormState`
+
+Note the difference in the server action signature when using `useFormState`:
+
+```ts
+export async function deleteSomethingWithResponse(
+  _prevState: any,
+  params: FormData
+) {
+  // ...
+}
+```
+
+as opposed to
+
+```ts
+export async function deleteSomething(params: FormData) {
+  // ...
+}
+```
+
+When using `useFormState` the server action receives a `prevState` parameter as the first argument. If you get an error like this:
+
+```ts
+TypeError: params.get is not a function
+```
+
+It's likely because you are using `useFormState` and your server action is not expecting the `prevState` parameter.
+
+## Conclusion
+
+This article has presented the basics of server actions, and how to use them in combination with actions that require user confirmation. Hopefully you found these examples useful and can apply them to your own projects. Next.js 13 allows us to use server actions, but they are still experimental, so expect possible breaking changes and improvements in the future.
```

Comment:
This is a really strong conclusion! I recommend more active voice wherever possible. For example, "This article presents the basics of server actions in Next.js, and how to use them with actions that require user confirmation. Next.js 13 introduces _experimental_ server actions, so certainly expect breaking changes and improvements in the future."